presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java

```diff
@@ -190,11 +193,21 @@ public Plan plan(Analysis analysis, Stage stage)
 
         planChecker.validateIntermediatePlan(root, session, metadata, sqlParser, variableAllocator.getTypes(), warningCollector);
         boolean enableVerboseRuntimeStats = SystemSessionProperties.isVerboseRuntimeStatsEnabled(session);
+        ImmutableList.Builder<PlanOptimizerInformation> optimizerOutput = ImmutableList.builder();
         if (stage.ordinal() >= Stage.OPTIMIZED.ordinal()) {
             for (PlanOptimizer optimizer : planOptimizers) {
                 long start = System.nanoTime();
-                root = optimizer.optimize(root, session, variableAllocator.getTypes(), variableAllocator, idAllocator, warningCollector);
+                PlanOptimizerOutput planOptimizerOutput = optimizer.optimize(root, session, variableAllocator.getTypes(), variableAllocator, idAllocator, warningCollector);
+                root = planOptimizerOutput.getPlanNode();
                 requireNonNull(root, format("%s returned a null plan", optimizer.getClass().getName()));
+                // Log only if the optimizer is triggered or could trigger but is disabled
+                if (planOptimizerOutput.getOptimizerInformation().isPresent()) {
+                    PlanOptimizerInformation optimizerInformation = planOptimizerOutput.getOptimizerInformation().get();
+                    if (optimizerInformation.getOptimizerTriggered()
+                            || (optimizerInformation.getOptimizerApplicable().isPresent() && optimizerInformation.getOptimizerApplicable().get())) {
```

Comment:
There shouldn't be anything else?Looks like this check is unnecessary? Maybe just log unconditionally?

Reply:
For an optimizer rule, it can have four status:
1. Enabled, and triggered
2. Enabled, and not triggered
3. Disabled, and could be triggered if enabled
4. Disabled, and could not be triggered.

I was planning to log only for case 1 and 3. Which is the check here.

And I am using an optional for the optimizerApplicable field, where Optional.empty() corresponds to the case where the rule does not return this information (it's disabled, and the optimizer also does not return information on whether it could be triggered if enabled).

To populate this field, we will need complex code change for each optimization rule (as we need to be able to do a dry run for the optimizer). I assume that we are not to implement this function for all existing rules, but only rules which we are interested in. Hence optional.empty is the default for this field.


> Maybe just log unconditionally?

Will this introduce too much overhead on the logging?